test: Fix 'make cov' with clang

hebasto opened the pull request:
This is a follow up of #19688.

With this PR it is possible to do the following:
```
$ ./autogen.sh
$ ./configure --enable-lcov CC=clang CXX=clang++
$ make
$ make cov
```

Currently, on master (8a85377cd0b60cb00dae4f595d628d1afbd28bd5), `make cov` fails to `Processing src/test/test_bitcoin-util_tests.gcda`.

hebasto committed:
test: Fix 'make cov' with clang

practicalswift commented:
@hebasto What is the root cause of this issue? :)

hebasto commented:
@practicalswift 
> @hebasto What is the root cause of this issue? :)

I don't know ((

I can just describe a pattern on which Boost Test suite and Clang's `llvm-cov` are incompatible: `BOOST_CHECK()` macro parameter calls some function more then once.

laanwj commented:
> @hebasto What is the root cause of this issue? :)

That's pretty important to figure out imo, we can't just randomly perturb the test code to make a tool happy.

fjahr commented:
Agree we should know what is going on but since I have had this issue locally I have gone ahead and tested it. So, I am tACK this PR cherry-picked on top of #19688 modulo an explanation of why this failed previously. Thanks for fixing this!

hebasto commented:
@practicalswift @laanwj @fjahr 
> What is the root cause of this issue? :)

Thank you for forcing me to find the root of this issue!

[`BOOST_TEST` - Limitations & workaround](https://www.boost.org/doc/libs/1_73_0/libs/test/doc/html/boost_test/testing_tools/boost_test_universal_macro.html#boost_test.testing_tools.boost_test_universal_macro.limitations_workaround):
> ... compound statements containing any logical composition `||`, `&&`... should be surrounded by parenthesis

To solve this issues some variants of the patch are available:

- as suggested by Boost docs:
```diff
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -559,11 +559,10 @@ BOOST_AUTO_TEST_CASE(util_ReadConfigStream)
                 && test_args.m_settings.ro_config[""].count("h")
                 && test_args.m_settings.ro_config[""].count("i")
                );
-    BOOST_CHECK(test_args.m_settings.ro_config["sec1"].count("ccc")
-                && test_args.m_settings.ro_config["sec1"].count("h")
-                && test_args.m_settings.ro_config["sec2"].count("ccc")
-                && test_args.m_settings.ro_config["sec2"].count("iii")
+    BOOST_CHECK(((test_args.m_settings.ro_config["sec1"].count("ccc")
+                && test_args.m_settings.ro_config["sec1"].count("h"))
+                && test_args.m_settings.ro_config["sec2"].count("ccc"))
+                && test_args.m_settings.ro_config["sec2"].count("iii")
                );
 
     BOOST_CHECK(test_args.IsArgSet("-a")
                 && test_args.IsArgSet("-b")
```

- using a local variable:
```diff
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -559,11 +559,10 @@ BOOST_AUTO_TEST_CASE(util_ReadConfigStream)
                 && test_args.m_settings.ro_config[""].count("h")
                 && test_args.m_settings.ro_config[""].count("i")
                );
-    BOOST_CHECK(test_args.m_settings.ro_config["sec1"].count("ccc")
-                && test_args.m_settings.ro_config["sec1"].count("h")
-                && test_args.m_settings.ro_config["sec2"].count("ccc")
-                && test_args.m_settings.ro_config["sec2"].count("iii")
-               );
+    bool test_value = test_args.m_settings.ro_config["sec1"].count("ccc")
+                && test_args.m_settings.ro_config["sec1"].count("h")
+                && test_args.m_settings.ro_config["sec2"].count("ccc")
+                && test_args.m_settings.ro_config["sec2"].count("iii");
+    BOOST_CHECK(test_value);
 
     BOOST_CHECK(test_args.IsArgSet("-a")
                 && test_args.IsArgSet("-b")
```

- as suggested in this PR:
```diff
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -559,11 +559,10 @@ BOOST_AUTO_TEST_CASE(util_ReadConfigStream)
                 && test_args.m_settings.ro_config[""].count("h")
                 && test_args.m_settings.ro_config[""].count("i")
                );
-    BOOST_CHECK(test_args.m_settings.ro_config["sec1"].count("ccc")
-                && test_args.m_settings.ro_config["sec1"].count("h")
-                && test_args.m_settings.ro_config["sec2"].count("ccc")
-                && test_args.m_settings.ro_config["sec2"].count("iii")
-               );
+    BOOST_CHECK(test_args.m_settings.ro_config["sec1"].count("ccc"));
+    BOOST_CHECK(test_args.m_settings.ro_config["sec1"].count("h"));
+    BOOST_CHECK(test_args.m_settings.ro_config["sec2"].count("ccc"));
+    BOOST_CHECK(test_args.m_settings.ro_config["sec2"].count("iii"));                );
 
     BOOST_CHECK(test_args.IsArgSet("-a")
                 && test_args.IsArgSet("-b")
```

The last variant has the following benefits:
- no error-prone parenthesis paring
- no redundant local variables
- better logging

Crypt-iQ commented:
Tested, code review ACK 568d42b

Changes are straightforward and they assert the same thing as they did before so should be fine.

test_bitcoin coverage: https://crypt-iq.github.io/pr19709/test_bitcoin.coverage/
total coverage: https://crypt-iq.github.io/pr19709/total.coverage/

maflcko reviewed (commented):
Approach ACK. This will also print nicer error messages in case of failure

vasild commented on src/test/util_tests.cpp:
```diff
@@ -550,57 +550,52 @@ BOOST_AUTO_TEST_CASE(util_ReadConfigStream)
+    BOOST_CHECK(!test_args.IsArgSet("-zzz"));
+    BOOST_CHECK(!test_args.IsArgSet("-iii"));
+
+    BOOST_CHECK(test_args.GetArg("-a", "xxx") == "");
```
```suggestion
    BOOST_CHECK_EQUAL(test_args.GetArg("-a", "xxx"), "");
```

would produce a better error message, if it fails some day.

Same in all places below where `BOOST_CHECK(a == b);` can be replaced with `BOOST_CHECK_EQUAL(a, b);`

hebasto replied:
Thank you! [Updated](https://github.com/bitcoin/bitcoin/pull/19709#issuecomment-673983005).

vasild reviewed (approved):
ACK 568d42bd7

It is always better to have
```
assert(a);
assert(b);
```
instead of
```
assert(a && b);
```
due to better diagnostics in case of failure.

The patch is good as is, but if you are in a mood for it - take the suggestion below.

hebasto commented:
Updated 568d42bd7f47851def00b185c6f5f285b6f7f951 -> 35cd2da623e32b975fbc485c3605934e4aa8bdc5 ([pr19709.01](https://github.com/hebasto/bitcoin/commits/pr19709.01) -> [pr19709.02](https://github.com/hebasto/bitcoin/commits/pr19709.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19709.01..pr19709.02)):

- addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/19709#discussion_r470460457):
> Same in all places below where `BOOST_CHECK(a == b);` can be replaced with `BOOST_CHECK_EQUAL(a, b);`

vasild commented:
ACK 35cd2da

Crypt-iQ commented:
ACK 35cd2da
